### ISSUE TITLE
Flycheck isn't reporting errors while a custom-checker is reporting them in checkstyle output

### ISSUE THREAD

**Issue by channikhabra** [flycheck/flycheck#810]
I've written a simple [custom checker](https://github.com/channikhabra/flycheck-canonical/blob/master/flycheck-canonical-scss.el) for a [personal linter tool](https://github.com/gajus/canonical) that we use. I am facing a weird issue that the linter gives error in checkstyle output, but flycheck doesn't report it.

I disable other checkers on opening the project file, and explicitly enable flycheck with:

```elisp
(progn
        (setq-local flycheck-disabled-checkers (append flycheck-disabled-checkers
                                                       '(scss-lint css-csslint scss sass)))
        (flycheck-select-checker 'canonical-scss)
        (message "Setup canonical-scss!"))
```

While in the scss buffer, when I run the `flycheck-compile` for canonical-scss, it shows the errors:

```xml
-*- mode: compilation; default-directory: "~/Documents/Work/xxx/yyy/src/static/styles/components/app/" -*-
Compilation started at Fri Nov 27 10:35:01

/Users/channi/Documents/Work/XXX/canonical/dist/bin/index.js --stdin --linter scss --outputFormat checkstyle --filePath /Users/channi/Documents/Work/XXX/YYY/src/static/styles/components/app/_map.scss < /Users/channi/Documents/Work/XXX/YYYY/src/static/styles/components/app/_map.scss
<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name="/Users/channi/Documents/Work/XXX/YYY/src/static/styles/components/app/_map.scss"><error line="1" column="6" severity="warning" source="" message="[property-sort-order] Expected "position, z-index, top, right, bottom, left, background", found "position, z-index, top, right, left, bottom, background"" /><error line="18" column="5" severity="warning" source="" message="[mixins-before-declarations] Mixins should come before declarations" /><error line="30" column="13" severity="warning" source="" message="[property-sort-order] Expected "position, z-index, top, left, display, width, height", found "position, z-index, top, left, display, height, width"" /></file></checkstyle>

Compilation finished at Fri Nov 27 10:35:02

```

But the same errors are not reported in the buffer itself. `flycheck-next-errors` say "No more Flycheck errors".

Do you have a clue what might be going wrong here?

**Comment by lunaryorn** [comment]
At a first glance your XML output appears to be invalid; there are quotes within quoted attribute values in the `property-sort-order` message.  Can you please run the command on your shell and pipe the output to `xmllint` to check whether it's valid XML?

**Comment by channikhabra** [comment]
That was stupid of me. I ran the xml through a random xml validator, and it said it's valid. Anyway, after escaping the quotes in the reported xml, some of the cases got fixed. However, the case whose report I posted above is still standing.

The exported xml gets a black space (` `) somehow in front of some of escaped quotes (`\"` become `\ "`) for some reason. This doesn't show up in the console. This happens only when I `flycheck-compile`:

```xml
-*- mode: compilation; default-directory: "~/Documents/Work/XXX/YYY/src/static/styles/components/app/" -*-
Compilation started at Fri Nov 27 14:33:26

/Users/channi/Documents/Work/XXX/canonical/dist/bin/index.js --stdin --linter scss --outputFormat checkstyle --filePath /Users/channi/Documents/Work/XXX/YYY/src/static/styles/components/app/_creative-primary.scss < /Users/channi/Documents/Work/XXX/YYY/src/static/styles/components/app/_creative-primary.scss
<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name="/Users/channi/Documents/Work/XXX/YYY/src/static/styles/components/app/_creative-primary.scss"><error line="1" column="19" severity="warning" source="" message="[property-sort-order] Expected  \"position, z-index, top, bottom, left, overflow, width, user-select, background, background-size \", found  \"position, top, z-index, left, bottom, overflow, width, user-select, background, background-size \" " /></file></checkstyle>

Compilation finished at Fri Nov 27 14:33:27
```

**Comment by channikhabra** [comment]
That whitespace didn't disappeared when I pasted the output here. Must be some other hidden char I guess? This is how the xml generated in `flycheck-compile` looks like when moved to a web-mode buffer:

![screen shot 2015-11-27 at 2 39 00 pm](https://cloud.githubusercontent.com/assets/1437091/11437551/a385910a-9514-11e5-9fe3-9e88a8edd7c9.png)

**Comment by channikhabra** [comment]
Damn I am on some kind of stupid streak today. Those whitespaces are added by the html-formatter (web-beautify-html), when I don't beautify it, there are no whitespaces in escaped quotes. But flycheck is not reporting these errors.

**Comment by lunaryorn** [comment]
@channikhabra That is _still_ invalid XML; please [escape your quotes correctly](http://stackoverflow.com/questions/1222367/escape-quote-character-in-xml).

With all due respect, please do try to make sure that your linter works correctly.  I'm closing this issue as invalid.

**Comment by channikhabra** [comment]
Thanks for your help. And really sorry for being an idiot today 😓

**Comment by lunaryorn** [comment]
@channikhabra You're welcome, no harm done :relaxed:

Just being curious, but why don't you use an XML library to generate the checkstyle output?  Any proper library should take care of this…

**Comment by channikhabra** [comment]
> Just being curious, but why don't you use an XML library to generate the checkstyle output?

I was looking for how other jslinters export their reports in checkstyle, and saw the [eslint-checkstyle](https://github.com/iancmyers/eslint/blob/5654e8eef09fd85d3c7a73b7392f209c50a0b241/lib/formatters/checkstyle.js) formatter. This is similar to how they are doing it. So I did it the same way.

**Comment by lunaryorn** [comment]
@channikhabra The [current version of this formatter](https://github.com/eslint/eslint/blob/master/lib/formatters/checkstyle.js) escapes XML properly.

**Comment by channikhabra** [comment]
Thanks a lot. I should be more careful.
